apps/gdaltindex_lib.cpp

```diff
@@ -546,95 +559,112 @@ GDALDatasetH GDALTileIndex(const char *pszDest, int nSrcCount,
+                             "-lyr_name must be specified.");
+                    if (pbUsageError)
+                        *pbUsageError = true;
+                    return nullptr;
                 }
-                osFormat = aoDrivers[0];
+                CPLPushErrorHandler(CPLQuietErrorHandler);
+                poLayer = poTileIndexDSUnique->GetLayerByName(
+                    psOptions->osIndexLayerName.c_str());
+                CPLPopErrorHandler();
             }
         }
         else
         {
-            osFormat = psOptions->osFormat;
-        }
-        if (!EQUAL(osFormat.c_str(), "ESRI Shapefile"))
-            nMaxFieldSize = 0;
+            if (psOptions->osFormat.empty())
+            {
+                const auto aoDrivers =
+                    GetOutputDriversFor(pszDest, GDAL_OF_VECTOR);
+                if (aoDrivers.empty())
+                {
+                    CPLError(CE_Failure, CPLE_AppDefined,
+                             "Cannot guess driver for %s", pszDest);
+                    return nullptr;
+                }
+                else
+                {
+                    if (aoDrivers.size() > 1)
+                    {
+                        CPLError(
+                            CE_Warning, CPLE_AppDefined,
+                            "Several drivers matching %s extension. Using %s",
+                            CPLGetExtensionSafe(pszDest).c_str(),
+                            aoDrivers[0].c_str());
+                    }
+                    osFormat = aoDrivers[0];
+                }
+            }
+            else
+            {
+                osFormat = psOptions->osFormat;
+            }
 
-        auto poDriver =
-            GetGDALDriverManager()->GetDriverByName(osFormat.c_str());
-        if (poDriver == nullptr)
-        {
-            CPLError(CE_Warning, CPLE_AppDefined, "%s driver not available.",
-                     osFormat.c_str());
-            return nullptr;
+            auto poDriver =
+                GetGDALDriverManager()->GetDriverByName(osFormat.c_str());
+            if (poDriver == nullptr)
+            {
+                CPLError(CE_Warning, CPLE_AppDefined,
+                         "%s driver not available.", osFormat.c_str());
+                return nullptr;
+            }
+
+            poTileIndexDSUnique.reset(
+                poDriver->Create(pszDest, 0, 0, 0, GDT_Unknown, nullptr));
+            if (!poTileIndexDSUnique)
+                return nullptr;
         }
 
-        poTileIndexDS.reset(
-            poDriver->Create(pszDest, 0, 0, 0, GDT_Unknown, nullptr));
-        if (!poTileIndexDS)
-            return nullptr;
+        poTileIndexDS = poTileIndexDSUnique.get();
+    }
+
+    if (osFormat.empty())
+    {
+        if (auto poOutDrv = poTileIndexDS->GetDriver())
+            osFormat = poOutDrv->GetDescription();
     }
 
+    const int nMaxFieldSize =
+        EQUAL(osFormat.c_str(), "ESRI Shapefile") ? 254 : 0;
```

Comment:
Maybe out of scope for this PR but can we avoid this kind of hacks? Perhaps a driver metadata that can be queried to know the field size limit?

Reply:
shapefile is pretty much  the only exception I can think of